sweep/tx_input_set.go

```diff
@@ -561,30 +561,43 @@ var _ InputSet = (*BudgetInputSet)(nil)
 
 // validateInputs is used when creating new BudgetInputSet to ensure there are
 // no duplicate inputs and they all share the same deadline heights, if set.
-func validateInputs(inputs []SweeperInput) error {
+func validateInputs(inputs []SweeperInput, deadlineHeight int32) error {
 	// Sanity check the input slice to ensure it's non-empty.
 	if len(inputs) == 0 {
 		return fmt.Errorf("inputs slice is empty")
 	}
 
-	// dedupInputs is a map used to track unique outpoints of the inputs.
-	dedupInputs := make(map[wire.OutPoint]struct{})
-
-	// deadlineSet stores unique deadline heights.
-	deadlineSet := make(map[fn.Option[int32]]struct{})
-
-	for _, input := range inputs {
-		input.params.DeadlineHeight.WhenSome(func(h int32) {
-			deadlineSet[input.params.DeadlineHeight] = struct{}{}
-		})
-
-		dedupInputs[input.OutPoint()] = struct{}{}
-	}
+	// inputDeadline tracks the input's deadline height. It will be updated
+	// if the input has a different deadline than the specified
+	// deadlineHeight.
+	inputDeadline := deadlineHeight
+
+	// dedupInputs is a set used to track unique outpoints of the inputs.
+	dedupInputs := fn.NewSet(
+		// Iterate all the inputs and map the function.
+		fn.Map(func(inp SweeperInput) wire.OutPoint {
+			// If the input has a deadline height, we'll check if
+			// it's the same as the specified.
+			inp.params.DeadlineHeight.WhenSome(func(h int32) {
+				// Exit early if the deadlines matched.
+				if h == deadlineHeight {
+					return
+				}
+
+				// Update the deadline height if it's
+				// different.
+				inputDeadline = h
+			})
```

Comment:
You mean creating another set of deadline heights and find the single different value using `Any`?

Reply:
Yeah rather than relying on mutating `inputDeadline` and other side-effecting behavior, we can use more functional computations to clean things up. The value of this is that it encourages us to think at higher level operations rather than deducing the memory changing behavior. It won't always be possible in go but in this case it's pretty straightforward so I'm suggesting doing so.

Reply:
I don't think this is better,
```diff
diff --git a/sweep/tx_input_set.go b/sweep/tx_input_set.go
index a4b4a746c..23d77a59c 100644
--- a/sweep/tx_input_set.go
+++ b/sweep/tx_input_set.go
@@ -567,10 +567,7 @@ func validateInputs(inputs []SweeperInput, deadlineHeight int32) error {
 		return fmt.Errorf("inputs slice is empty")
 	}
 
-	// inputDeadline tracks the input's deadline height. It will be updated
-	// if the input has a different deadline than the specified
-	// deadlineHeight.
-	inputDeadline := deadlineHeight
+	deadlineHeights := make([]int32, 0, len(inputs))
 
 	// dedupInputs is a set used to track unique outpoints of the inputs.
 	dedupInputs := fn.NewSet(
@@ -579,23 +576,30 @@ func validateInputs(inputs []SweeperInput, deadlineHeight int32) error {
 			// If the input has a deadline height, we'll check if
 			// it's the same as the specified.
 			inp.params.DeadlineHeight.WhenSome(func(h int32) {
-				// Exit early if the deadlines matched.
-				if h == deadlineHeight {
-					return
-				}
-
-				// Update the deadline height if it's
-				// different.
-				inputDeadline = h
+				deadlineHeights = append(deadlineHeights, h)
 			})
 
 			return inp.OutPoint()
 		}, inputs)...,
 	)
 
+	// inputDeadline tracks the input's deadline height. It will be updated
+	// if the input has a different deadline than the specified
+	// deadlineHeight.
+	var inputDeadline int32
+
+	different := fn.Any(func(h int32) bool {
+		if h == deadlineHeight {
+			return true
+		}
+
+		inputDeadline = h
+		return false
+	}, deadlineHeights)
+
 	// Make sure the inputs share the same deadline height when there is
 	// one.
-	if inputDeadline != deadlineHeight {
+	if different {
 		return fmt.Errorf("input deadline height not matched: want "+
 			"%d, got %d", deadlineHeight, inputDeadline)
 	}
```

What's more important is I don't think `Any` works - it will return false when the supplied slice is empty, which can happen here when inputs all have deadline none - unless I use `noDifferent` and check `h != deadlineHeight`, then it's more error-prone.

I think I can convince myself with the design of `Any` re the empty slice behavior - since no item satisfy the predicate, we should return false. But then it will conflict with `All`. Why do we use `if pred` in one place and `if !pred` in the other?

Reply:
The functions are dual to each other. It has to do with their folding behavior.

```haskell
all = foldr (&&) True
any = foldr (||) False
```
The idea is that you are doing successive combining with the accumulator.

I think you revealed the solution though, just use `All`  for the negated condition.

If you strongly oppose it, then you can proceed, I just am strongly of the opinion that the less we have to reason about mutation the better. Declarative code is easier to understand as it doesn't require you to think about *when* things happen in order to understand *what* is happening